### PR TITLE
daml2js: Write all dependencies into package.json

### PR DESCRIPTION
**Pull request by hurryabit**
The dependencies of the generated packages on `@daml/types` and
`@mojotech/json-type-validation` somehow got lost either during a
refactoring or when changing how we depend on other generated packages.

The tests simply running `daml2ts` did not catch this because we write
a temporary `package.json` one level higher up in the directory tree
in order to `yarn install` the dependencies only once.

The tests using the generated packages also did not catch this since
they depend on `@daml/types`, which in turn depends on
`@mojotech/json-type-validation`, themselves and hence had both packages
in scope.

In other words, this is very tricky to test. We try to alleviate this
problem by sharing the code for generating the dependencies between the
function writing the actual `package.json` files and the one writing
the temporary `package.json`. This makes sure that these dependencies
do not get out of sync again.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/5490)

### CODE REVIEW

**Comment by hurryabit** [comment]
It wasn't a problem because the `package.json` in `ui` depends on `@daml/types` directly (and hence on `@mojotech/json-type-validation` indirectly). Thus, running `yarn install` puts these packages into `ui/node_modules`. The `daml codegen js ...` part succeeded because of the temporary `package.json` file we write. Does that explain why it worked?

**Comment by cocreature** [comment]
Makes sense, thanks for the explanation!

**Comment by cocreature** [comment]
CI failures seem legit.

File: language-support/ts/codegen/tests/build-and-lint.sh [hunk 1]
```diff
@@ -60,16 +60,10 @@ cd $TMP_DIR
 
 # Call daml2js.
 PATH=`dirname $YARN`:$PATH $DAML2TS -o daml2js $DAR
-# We only needed this package.json in order to provide resolution of
-# daml-types to daml2js and that's done with now.
-rm package.json
-# Since we eschew using workspaces and resolutions in this test, patch
-# up the daml-ledger dependency on daml-types.
-sed -i"" "s|\"@daml/types\": \"$SDK_VERSION\"|\"@daml/types\": \"file:../daml-types\"|" daml-ledger/package.json
```
Comment (hurryabit) [inline language-support/ts/codegen/tests/build-and-lint.sh:68]: This really needs to become a proper Haskell test. I'll take care of that once this PR has landed.

**Comment by hurryabit** [comment]
@cocreature Yes, they were legit indeed. I've pushed a fix in the meantime. The huge bash script is slowly turning into a giant pain. Calls for a proper test suite in Haskell. 😸

**Pull request merged**